fix: Explicitly exit CLI to prevent hanging subprocesses

veracioux opened the pull request:
Reproduced using MCP config from: https://github.com/sst/opencode/issues/1717

The easiest way to reproduce (and verify the fix) is to run `opencode run hi`.

Investigation:
- The MCP servers from the issue are run via `docker run`
- When opencode exits, the `ai` library sends SIGTERM to all child MCP processes that use stdio transports
- The problem is that the child processes are the `docker run` client, and `SIGTERM` isn't properly delivered, unless `docker run --init` was used. So the subprocesses stick around, preventing opencode from exiting. I verified this with a minimalistic subprocess-spawning bun program and `strace`
- What does work though, is killing the opencode process itself using `process.exit`, in which case `SIGTERM` is delivered correctly
- Fix: I added a `process.exit()` when the CLI is done with everything.

NOTE: Issue also happens in (some, maybe all) MCP servers wrapped by shell scripts, which this also fixes.

---

Edit:
I also found some places that don't properly await Promises and are liable to cause hangs before process exit. I changed those. I also replaced some `for () { await ... }` with `await Promise.all(...)` to potentially speed up cleanups of unrelated resources (which hence can be performed concurrently).

Improved INFO logs for ensureTitle - it's now awaited and logged so it can be more obvious from `--print-logs` that that is the reason why OC hangs before exiting.

---

Closes #854.
Closes #1001.
Closes #1195.
Closes #1431.
Closes #1717.
Closes #1810.

Might fix:
- #2406.
- #3213

Possibly related, but I'm not very confident this is fixed by the PR:
- #2819

veracioux committed:
fix: Explicitly exit process to prevent hanging subprocesses

veracioux committed:
fix: Explicitly exit process to prevent hanging subprocesses

veracioux committed:
refactor: Remove unused `cancel` AbortController

veracioux committed:
refactor: Remove unused `cancel` AbortController

veracioux commented:
Rebased to remove unintended extra commit.

veracioux committed:
refactor: Improve promise handling

veracioux committed:
refactor: Improve promise handling

veracioux committed:
fix: Await ensureTitle before app exits

veracioux committed:
fix: Await ensureTitle before app exits

rekram1-node commented on packages/opencode/src/session/prompt.ts:
```diff
@@ -335,6 +336,9 @@ export namespace SessionPrompt {
       state().queued.delete(input.sessionID)
       SessionCompaction.prune(input)
+      log.info("waiting for session title to be generated")
+      await ensureTitlePromise
```
are you sure we wanna await here, doesnt that make title generation blocking

veracioux replied:
Yes, definitely. The promise will be running either way, and if we don't await it here, it will hang around until either it resolves or the process exits.

I think it makes sense to await it here, as it is the end of the LLM's response to the user's input, meaning it can run concurrently with the LLM's response.

Also, since I do `process.exit()` at the end of `index.ts` to work around those dockerized MCP servers, if I don't await this here, the `process.exit()` will be reached before the ensureTitlePromise resolves, so it will be interrupted and the title won't be saved.

veracioux replied:
@rekram1-node Tweaked slightly to address your concern. See https://github.com/sst/opencode/pull/3083/commits/1e0846fb63fb25c09f7b1839fb7a360b2a755ec6.

veracioux committed:
feat: Move `await ensureTitlePromise` to state disposal

veracioux committed:
feat: Move `await ensureTitlePromise` to state disposal

rekram1-node commented on packages/opencode/src/index.ts:
```diff
@@ -129,6 +127,10 @@ try {
+  // Most notably, some docker-container-based MCP servers don't handle such signals unless
+  // run using `docker run --init`.
+  // Explicitly exit to avoid any hanging subprocesses.
+  process.exit();
```
are there any other cases besides the docker mcps?

Can't we just track the pid from transports and send a SIGKILL if they are still running post close?

veracioux replied:
I have a shell script wrapping an stdio MCP that has the same behavior. I can send it to you later today. I can imagine there can be other cases where an MCP misbehaves in a similar way. I have seen a few MCP server implementations and the general vibe is not high quality or well thought out.

I would like to avoid using SIGKILL. Fact of the matter is, the actual MCP server process receives SIGTERM when process.exit() is called. I confirmed that by running `strace` on the container's PID. But from the POV of OC, the child process is not the MCP server, but rather the docker client managing the MCP server.

The PID from the transports is an implementation detail - I would like to avoid using that too.

Any subprocess not under our control can cause the hang. The process.exit is a panacea, and as long as we do proper awaits, it won't cause any issues. Which we do for the most part.

veracioux replied:
The shell script + how I use it in `opencode.json`, as promised: https://gist.github.com/veracioux/3a9fe2ea06fca0d16852481030d9297e

rekram1-node commented on packages/opencode/src/session/prompt.ts:
```diff
@@ -76,10 +76,14 @@ export namespace SessionPrompt {
 
       return {
         queued,
+        ensureTitlePromise: undefined as Promise<void> | undefined,
```
what if I have multiple calls to prompt at once?

rekram1-node replied:
Keep in mind that opencode server isn't tied just to the normal server + tui, it can also have other use cases ex:

https://github.com/sst/opencode/blob/dev/packages/sdk/js/example/example.ts

veracioux replied:
Ah yes, I'll make some tweaks. Thanks for pointing out.

veracioux replied:
See newest commits. They should address both these concerns.

veracioux committed:
feat: Improve logs and robustness of state disposal

- Track cleanup promises (LSP clients, MCP clients, file watchers) to
  prevent premature disposal
- Add timeout warnings to alert users when disposal is taking longer
  than expected
- Use allSettled instead of Promise.all to handle disposal errors gracefully
- Move session title generation to tracked promises, let State.dispose
  handle it in unified way

veracioux committed:
feat: Improve logs and robustness of state disposal

- Track cleanup promises (LSP clients, MCP clients, file watchers) to
  prevent premature disposal
- Add timeout warnings to alert users when disposal is taking longer
  than expected
- Use allSettled instead of Promise.all to handle disposal errors gracefully
- Move session title generation to tracked promises, let State.dispose
  handle it in unified way

veracioux committed:
feat: Extend disposal timeout and improve log message

veracioux committed:
feat: Extend disposal timeout and improve log message

rekram1-node commented on packages/opencode/src/file/watcher.ts:
```diff
@@ -63,7 +63,8 @@ export namespace FileWatcher {
     async (state) => {
-      state.sub?.unsubscribe()
+      if (!state.sub) return
+      Instance.trackPromises([state.sub.unsubscribe()])
```
why do we need to trackPromises it if this is already in the dispose function? Cant we just await it?

rekram1-node committed:
some tweaks

rekram1-node committed:
Merge branch 'exit-tweaks' into bugfix/opencode-hangs-after-exit

rekram1-node committed:
fix

rekram1-node committed:
cleanup
